### ISSUE TITLE
[Feature] Submission & XP System

### ISSUE THREAD

**Issue by nauvalazhar** [nauvalazhar/code-design#7]
Let's have the submission page and the XP system to make the challenge more engaging.

Submission is a page that would allow the dev to upload their challenge's solution. Each submission contains at least a description, demo URL, and tech stack information.

All the submissions would go through the approval system, and after the submission is approved, the dev would earn some XPs.

I still don't have any idea about the tech that we can use, but I think the GitHub repository's issues API would be the easier way to implement. But, we still need to find a way to store the user's XP data.

I don't want to build an entire back-end infrastructure only for this feature to be implemented.

Some other advanced options:
- https://planetscale.com
- https://strapi.io

Let's have a discussion.

**Comment by hendraaagil** [comment]
I think if want to make an XP system, we need to build the our own back-end application 😅
Because I can't find any platform who can make that system.

With custom back-end app, maybe we can also make people can submit the design asset through the website.

**Comment by nauvalazhar** [comment]
I see, my concern is about the deployment, what option for the deployment, then? I think we can do a full-stack Next JS with the Planetscale database, they have got pretty good a free plan.

**Comment by hendraaagil** [comment]
I think full-stack NextJS with Planetscale database is a good idea.

**Comment by akrindev** [comment]
does it also mean to have badges system?

**Comment by nauvalazhar** [comment]
> does it also mean to have badges system?

Yeah, it is on my todo-list. But, for now, let's have the submission page first, so the dev could publish the solutions and share them with the world!

**Comment by akrindev** [comment]
> 

sounds good✨
then, do u think we need authentication system? wbu to use github provider? the user then have their submission profile, xp and badges histories.

here is how I figured out
<img width="3296" alt="codedesign flow" src="https://user-images.githubusercontent.com/33445459/218441991-9c3cad74-6e8b-4253-ab15-0079ac0f3210.png">
I'm so excited to collaborate 🥳

**Comment by nauvalazhar** [comment]
Whoa, such a nice diagram! 

Yes, we can use GitHub provider, but we still need to find a way to persist the user login information on the client. I've tried to implement the session-based auth system, but the problem is we use Next JS 13 appDir feature, so the session API is still read-only. But we can use JWT as an alternative, or maybe you have another option.

**Comment by akrindev** [comment]
> Whoa, such a nice diagram!
> 
> Yes, we can use GitHub provider, but we still need to find a way to persist the user login information on the client. I've tried to implement the session-based auth system, but the problem is we use Next JS 13 appDir feature, so the session API is still read-only. But we can use JWT as an alternative, or maybe you have another option.

its a great idea to use JWT, I've been using next-auth to handle authentication system with appDir. I'll go first with PR login system.

**Comment by nauvalazhar** [comment]
> > Whoa, such a nice diagram!
> > Yes, we can use GitHub provider, but we still need to find a way to persist the user login information on the client. I've tried to implement the session-based auth system, but the problem is we use Next JS 13 appDir feature, so the session API is still read-only. But we can use JWT as an alternative, or maybe you have another option.
> 
> its a great idea to use JWT, I've been using next-auth to handle authentication system with appDir. I'll go first with PR login system.

Oh, sounds good!

**Comment by rndy28** [comment]
Hi, new here, I suggest using trpc for custom backend and we can easily integrate it with next auth also, like [t3-stack](https://github.com/t3-oss/create-t3-app)

> I think if want to make an XP system, we need to build the our own back-end application 😅 Because I can't find any platform who can make that system.
> 
> With custom back-end app, maybe we can also make people can submit the design asset through the website.

**Comment by fzn0x** [comment]
> I think if want to make an XP system, we need to build the our own back-end application 😅 Because I can't find any platform who can make that system.
> 
> With custom back-end app, maybe we can also make people can submit the design asset through the website.

Alternatively, we can use https://strapi.io/

**Comment by fzn0x** [comment]
> I think if want to make an XP system, we need to build the our own back-end application 😅 Because I can't find any platform who can make that system.
> 
> With custom back-end app, maybe we can also make people can submit the design asset through the website.

Alternatively, we can use https://strapi.io/

For the point system as plugin: https://strapi.io/blog/how-to-build-a-strapi-custom-plugin
